### ISSUE TITLE
binReadCounts fails when setting chunksize

### ISSUE THREAD

**Issue by SebastianHollizeck** [ccagc/QDNAseq#89]
Hey,

As the title says, when i specify a chunksize to bin the reads, it fails with 

```
> readCounts <- binReadCounts(bins, bamfiles=bam, chunkSize = 1E7)
    BB0042_rmd (1 of 1): extracting reads ...
Error in getGlobalsAndPackages(expr, envir = envir, globals = globals) : 
  Did you mean to create the future within a function?  Invalid future expression tries to use global '...' variables that do not exist: FUN()
```

just the session info, because i assume it has to do with the future package 
```
> sessionInfo()
R version 4.0.2 (2020-06-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS:   /config/binaries/R/4.0.2/lib64/R/lib/libRblas.so
LAPACK: /config/binaries/R/4.0.2/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C               LC_TIME=en_AU.UTF-8       
 [4] LC_COLLATE=en_AU.UTF-8     LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
 [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] QDNAseq_1.26.0              data.table_1.13.6           VariantAnnotation_1.36.0   
 [4] Rsamtools_2.6.0             Biostrings_2.58.0           XVector_0.30.0             
 [7] SummarizedExperiment_1.20.0 Biobase_2.50.0              MatrixGenerics_1.2.1       
[10] matrixStats_0.58.0          GenomicRanges_1.42.0        GenomeInfoDb_1.26.2        
[13] IRanges_2.24.1              S4Vectors_0.28.1            BiocGenerics_0.36.0        

loaded via a namespace (and not attached):
 [1] bitops_1.0-6             bit64_4.0.5              RColorBrewer_1.1-2      
 [4] progress_1.2.2           httr_1.4.2               tools_4.0.2             
 [7] R6_2.5.0                 colorspace_2.0-0         DBI_1.1.1               
[10] DNAcopy_1.64.0           tidyselect_1.1.0         prettyunits_1.1.1       
[13] bit_4.0.4                curl_4.3                 compiler_4.0.2          
[16] VennDiagram_1.6.20       formatR_1.7              xml2_1.3.2              
[19] DelayedArray_0.16.1      rtracklayer_1.50.0       askpass_1.1             
[22] rappdirs_0.3.3           stringr_1.4.0            digest_0.6.27           
[25] rmarkdown_2.6            R.utils_2.10.1           pkgconfig_2.0.3         
[28] htmltools_0.5.1.1        parallelly_1.23.0        plotrix_3.8-1           
[31] dbplyr_2.0.0             fastmap_1.1.0            limma_3.46.0            
[34] BSgenome_1.58.0          rlang_0.4.10             rstudioapi_0.13         
[37] RSQLite_2.2.3            impute_1.64.0            generics_0.1.0          
[40] BiocParallel_1.24.1      dplyr_1.0.3              R.oo_1.24.0             
[43] RCurl_1.98-1.2           magrittr_2.0.1           GenomeInfoDbData_1.2.4  
[46] futile.logger_1.4.3      Matrix_1.3-2             Rcpp_1.0.6              
[49] ensemblVEP_1.32.0        lifecycle_0.2.0          R.methodsS3_1.8.1       
[52] yaml_2.2.1               stringi_1.5.3            CGHbase_1.50.0          
[55] MASS_7.3-53.1            zlibbioc_1.36.0          BiocFileCache_1.14.0    
[58] grid_4.0.2               blob_1.2.1               listenv_0.8.0           
[61] crayon_1.3.4             lattice_0.20-41          GenomicFeatures_1.42.1  
[64] hms_1.0.0                knitr_1.31               pillar_1.4.7            
[67] CGHcall_2.52.0           future.apply_1.7.0       codetools_0.2-18        
[70] marray_1.68.0            biomaRt_2.46.3           futile.options_1.0.1    
[73] XML_3.99-0.5             glue_1.4.2               evaluate_0.14           
[76] BiocManager_1.30.10      lambda.r_1.2.4           vctrs_0.3.6             
[79] openssl_1.4.3            purrr_0.3.4              future_1.21.0           
[82] assertthat_0.2.1         cachem_1.0.4             xfun_0.20               
[85] tibble_3.0.6             GenomicAlignments_1.26.0 tinytex_0.29            
[88] AnnotationDbi_1.52.0     memoise_2.0.0            corrplot_0.84           
[91] globals_0.14.0           ellipsis_0.3.1  
```

**Comment by HenrikBengtsson** [comment]
Thanks for reporting.  Could you please share `traceback()` after getting that error?

Also, just to confirm, are you saying it works fine if you do:

```r
readCounts <- binReadCounts(bins, bamfiles=bam)
```

?

**Comment by SebastianHollizeck** [comment]
Hey,

yes, i can confirm the function call works when not setting the chunk size.

here the output of the traceback
```
> readCounts <- binReadCounts(bins, bamfiles=bam, chunkSize = 1E7)
    BB0042_rmd (1 of 1): extracting reads ...
Error in getGlobalsAndPackages(expr, envir = envir, globals = globals) : 
  Did you mean to create the future within a function?  Invalid future expression tries to use global '...' variables that do not exist: FUN()
> traceback()
7: stop(msg)
6: getGlobalsAndPackages(expr, envir = envir, globals = globals)
5: getGlobalsAndPackagesXApply(FUN = FUN, args = args, MoreArgs = MoreArgs, 
       envir = envir, future.globals = future.globals, future.packages = future.packages, 
       debug = debug)
4: future_xapply(FUN = FUN, nX = nX, chunk_args = X, args = list(...), 
       get_chunk = `[`, expr = expr, envir = envir, future.globals = future.globals, 
       future.packages = future.packages, future.scheduling = future.scheduling, 
       future.chunk.size = future.chunk.size, future.stdout = future.stdout, 
       future.conditions = future.conditions, future.seed = future.seed, 
       future.lazy = future.lazy, future.label = future.label, fcn_name = fcn_name, 
       args_name = args_name, debug = debug)
3: future_lapply(names(targets), FUN = function(seqName) {
       readCounts <- integer(length = nrow(bins))
       seqNameI <- sub("chr", "", seqName)
       for (chunk in 1:ceiling(targets[seqName]/chunkSize)) {
           chunkStart <- (chunk - 1) * chunkSize + 1
           chunkEnd <- chunk * chunkSize + 1
           params <- ScanBamParam(flag = flag, what = c("rname", 
               "pos", "mapq"), which = GRanges(seqName, IRanges(chunkStart, 
               chunkEnd)))
           reads <- scanBam(bamfile, param = params)
           reads <- reads[[1L]]
           hasMapq <- any(is.finite(reads[["mapq"]]))
           if (hasMapq) {
               keep <- which(reads[["mapq"]] >= minMapq)
               reads <- lapply(reads, FUN = function(x) x[keep])
           }
           hits <- list()
           hits[[seqNameI]] <- reads[["pos"]]
           chunkName <- paste(seqName, chunkStart, chunkEnd, sep = ":")
           vmsg(paste("binning chunk -", chunkName, sep = " "), 
               appendLF = TRUE)
           keep <- which(bins$chromosome == seqNameI & bins$start >= 
               chunkStart & bins$end <= chunkEnd)
           if (length(keep) == 0L) 
               next
           chromosomeBreaks <- c(bins$start[keep], max(bins$end[keep]) + 
               1)
           counts <- binCounts(hits[[seqNameI]], bx = chromosomeBreaks)
           readCounts[keep] <- readCounts[keep] + counts
       }
       readCounts
   })
2: .binReadCountsPerChunk(bins = bins, bamfile = bamfiles[i], chunkSize = chunkSize, 
       cache = cache, force = force, isPaired = isPaired, isProperPair = isProperPair, 
       isUnmappedQuery = isUnmappedQuery, hasUnmappedMate = hasUnmappedMate, 
       isMinusStrand = isMinusStrand, isMateMinusStrand = isMateMinusStrand, 
       isFirstMateRead = isFirstMateRead, isSecondMateRead = isSecondMateRead, 
       isSecondaryAlignment = isSecondaryAlignment, isNotPassingQualityControls = isNotPassingQualityControls, 
       isDuplicate = isDuplicate, minMapq = minMapq)
1: binReadCounts(bins, bamfiles = bam, chunkSize = 1e+07)
```

**Comment by HenrikBengtsson** [comment]
Thanks.  I can reproduce with:

```r
library(QDNAseq)
library(QDNAseq.hg19)
bins <- getBinAnnotations(500, genome = "hg19")
bam <- system.file("extdata", "ex1.bam", package = "Rsamtools")

# Works
counts <- binReadCounts(bins, bamfiles = bam)
print(counts)

# Fails
counts <- binReadCounts(bins, bamfiles = bam, chunkSize = 10e6)
#     ex1 (1 of 1): extracting reads ...
# Error in getGlobalsAndPackages(expr, envir = envir, globals = globals) : 
#   Did you mean to create the future within a function?  Invalid future expression tries to use global '...' variables that do not exist: FUN()
```

**Comment by HenrikBengtsson** [comment]
This will most likely be fixed automatically when the next version (1.22.1) of **[future](https://cran.r-project.org/web/packages/future/index.html)** reaches CRAN.  It's in the CRAN queue right now.  Stay tuned.

**Comment by HenrikBengtsson** [comment]
I'm confirming that is fixed by updating to the latest **future** version (>= 1.22.1).  I've added package test for this and I pushed **QDNAseq** 1.29.2, which requires this version of **future**.
